test/core/security/grpc_tls_crl_provider_test.cc

```diff
@@ -83,6 +101,110 @@ TEST(CrlProviderTest, StaticCrlProviderLookupBad) {
   EXPECT_EQ(crl, nullptr);
 }
 
+TEST(CrlProviderTest, DirectoryReloaderCrlLookupGood) {
+  auto result = experimental::CreateDirectoryReloaderProvider(
+      kCrlDirectory, std::chrono::seconds(1), nullptr);
+  ASSERT_TRUE(result.ok());
+  std::shared_ptr<CrlProvider> provider = std::move(*result);
+
+  CertificateInfoImpl cert = CertificateInfoImpl(kCrlIssuer);
+  auto crl = provider->GetCrl(cert);
+  ASSERT_NE(crl, nullptr);
+  ASSERT_EQ(crl->Issuer(), kCrlIssuer);
+
+  CertificateInfoImpl intermediate =
+      CertificateInfoImpl(kCrlIntermediateIssuer);
+  auto intermediate_crl = provider->GetCrl(intermediate);
+  ASSERT_NE(intermediate_crl, nullptr);
+  ASSERT_EQ(intermediate_crl->Issuer(), kCrlIntermediateIssuer);
+}
+
+TEST(CrlProviderTest, DirectoryReloaderCrlLookupBad) {
+  auto result = experimental::CreateDirectoryReloaderProvider(
+      kCrlDirectory, std::chrono::seconds(1), nullptr);
+  ASSERT_TRUE(result.ok());
+  std::shared_ptr<CrlProvider> provider = std::move(*result);
+
+  CertificateInfoImpl bad_cert = CertificateInfoImpl("BAD CERT");
+  auto crl = provider->GetCrl(bad_cert);
+  ASSERT_EQ(crl, nullptr);
+}
+
+std::string MakeTempDir() {
+  char templ[] = "/tmp/tmpdir.XXXXXX";
+  std::string dir_path = mkdtemp(templ);
+  return dir_path;
+}
+
+std::string TempDirNameFromPath(absl::string_view dir_path) {
+  std::vector<std::string> split = absl::StrSplit(dir_path, "/");
+  return split[2] + "/";
+}
+
+TEST(CrlProviderTest, DirectoryReloaderReloadsAndDeletes) {
+  std::string dir_path = MakeTempDir();
+  std::string dir_name = TempDirNameFromPath(dir_path);
+
+  auto result = experimental::CreateDirectoryReloaderProvider(
+      dir_path, std::chrono::seconds(1), nullptr);
+  ASSERT_TRUE(result.ok());
+  std::shared_ptr<CrlProvider> provider = std::move(*result);
+  CertificateInfoImpl cert = CertificateInfoImpl(kCrlIssuer);
+  auto should_be_no_crl = provider->GetCrl(cert);
+  ASSERT_EQ(should_be_no_crl, nullptr);
+
+  {
+    std::string raw_crl = GetFileContents(kCrlPath);
+    TmpFile tmp_crl(raw_crl, dir_name);
```

Comment:
Instead of actually creating files here, how about a dependency-injection approach?  You can create an interface that wraps the directory reloading and have a fake version of that interface here in the test, so that you can control what the `DirectoryReloaderCrlProvider` sees each time it tries to read the directory.

Reply:
I think this will make the test a lot cleaner and help with portability - will take a stab at this after resolving smaller comments. Leaving this open for now

Reply:
PTAL